code/modules/vr/vr_avatar.dm

```diff
@@ -16,73 +17,128 @@
-	var/mob/living/carbon/human/H = real_me
-	revert_to_reality(FALSE)
-	if(H)
-		H.ghost()
-	qdel(src)
-
-/mob/living/carbon/human/virtual_reality/proc/revert_to_reality(remove = 0)
-	if(real_me && mind)
-		mind.transfer_to(real_me)
+	alert(src, "You can not ghost while in VR. Please exit first.","Are you sure you want to ghost?")
+	return
+
+/mob/living/carbon/human/virtual_reality/proc/revert_to_reality(var/remove)
+	if(real_me && ckey)
+		real_me.ckey = ckey
 		real_me.EyeBlind(2)
 		real_me.Confused(2)
+		var/mob/living/carbon/human/virtual_reality/vr = src
+		var/obj/item/clothing/ears/vr_headset/goggles
+		for(var/obj/item/clothing/ears/vr_headset/g in vr.real_me.contents)
+			goggles = g
+		if(goggles.contained() && vr_server_status == VR_SERVER_EMAG)
+			real_me.adjustBrainLoss(60)
+			real_me.vomit(20)
+		for(var/obj/item/clothing/ears/vr_headset/g in vr.real_me.contents)
+			processing_objects.Remove(g)
+			if(remove)
+				g.vr_human = null
 		if(remove)
-			qdel(src)
+			handle_despawn()
 		else
 			return src
 
 /mob/living/carbon/human/virtual_reality/proc/return_to_lobby()
 	if(real_me && mind)
 		var/mob/living/carbon/human/virtual_reality/new_vr
-		new_vr = spawn_vr_avatar(src, "Lobby")
-		var/obj/item/clothing/glasses/vr_goggles/g = new_vr.real_me.glasses
-		g.vr_human = new_vr
-		qdel(src)
+		var/datum/vr_room/lobby = vr_rooms_official["Lobby"]
+		new_vr = spawn_vr_avatar(src, lobby)
+		for(var/obj/item/clothing/ears/vr_headset/g in new_vr.real_me.contents)
+			g.vr_human = new_vr
 
 /datum/action/quit_vr
 	name = "Quit Virtual Reality"
+	button_icon_state = "shutdown"
 
 /datum/action/quit_vr/Trigger()
 	if(..())
 		if(istype(owner, /mob/living/carbon/human/virtual_reality))
 			var/mob/living/carbon/human/virtual_reality/vr = owner
-			vr.revert_to_reality(1)
+			var/obj/item/clothing/ears/vr_headset/goggles
+			for(var/obj/item/clothing/ears/vr_headset/g in vr.real_me.contents)
+				goggles = g
+			if(!goggles.contained())
+				vr.revert_to_reality(1)
+			else
+				to_chat(owner, "You are currently imprisoned in Virtual Reality and are unable to disconnect.")
 		else
 			Remove(owner)
 
 /datum/action/back_to_lobby
 	name = "Return To Lobby"
+	button_icon_state = "lobby"
 
 /datum/action/back_to_lobby/Trigger()
 	if(..())
 		if(istype(owner, /mob/living/carbon/human/virtual_reality))
 			var/mob/living/carbon/human/virtual_reality/vr = owner
-			vr.return_to_lobby()
+			vr.death()
 		else
 			Remove(owner)
 
 /datum/action/detach_from_avatar
 	name = "Detach From Avatar"
+	button_icon_state = "logout"
 
 /datum/action/detach_from_avatar/Trigger()
 	if(..())
 		if(istype(owner, /mob/living/carbon/human/virtual_reality))
 			var/mob/living/carbon/human/virtual_reality/vr = owner
-			var/obj/item/clothing/glasses/vr_goggles/g = vr.real_me.glasses
-			g.vr_human = vr
-			vr.revert_to_reality(0)
+			var/obj/item/clothing/ears/vr_headset/goggles
+			for(var/obj/item/clothing/ears/vr_headset/g in vr.real_me.contents)
+				goggles = g
```

Comment:
I feel like this is unnecessary as you already have a var for the VR goggles (`g`)

Reply:
g is in a loop though. I need a var outside a loop I think.